cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -560,10 +746,34 @@ public void deleteStream(String namespaceId, String name) {
     deleteAll(new MDSKey.Builder().add(TYPE_STREAM, namespaceId, name).build());
   }
 
-  public void deleteProgramHistory(String namespaceId, String appId) {
-    deleteAll(new MDSKey.Builder().add(TYPE_RUN_RECORD_STARTED, namespaceId, appId).build());
-    deleteAll(new MDSKey.Builder().add(TYPE_RUN_RECORD_COMPLETED, namespaceId, appId).build());
-    deleteAll(new MDSKey.Builder().add(TYPE_RUN_RECORD_SUSPENDED, namespaceId, appId).build());
+  public void deleteProgramHistory(String namespaceId, String appId, String versionId) {
+    Map<MDSKey, RunRecordMeta> runRecords;
+    if (versionId.equals(ApplicationId.DEFAULT_VERSION)) {
+      runRecords = listKV(new MDSKey.Builder().add(TYPE_RUN_RECORD_STARTED, namespaceId, appId).build(),
```

Comment:
This will give us the run records for all versions of the app, right? Instead of reading all the run records why don't we send a filter to deleteAll?